source/extensions/common/wasm/context.cc

```diff
@@ -0,0 +1,212 @@
+    break;
+  case proxy_wasm::LogLevel::error:
+    ENVOY_LOG(error, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::critical:
+    ENVOY_LOG(critical, "wasm log{}: {}", log_prefix(), message);
+    break;
+  default:
+    return WasmResult::BadArgument;
+  }
+  return WasmResult::Ok;
+}
+
+WasmResult Context::getProperty(absl::string_view path, std::string* result) {
+  if (path.size() == 15 && !memcmp(reinterpret_cast<const void*>(&*path.begin()),
+                                   reinterpret_cast<const void*>("plugin_root_id\0"), 15)) {
+    std::cerr << "returning " << root_id() << "\n";
+    *result = std::string(root_id());
+    return WasmResult::Ok;
+  }
+  return WasmResult::Unimplemented;
+}
+
+WasmResult Context::defineMetric(MetricType type, absl::string_view name, uint32_t* metric_id_ptr) {
+  auto stat_name = wasm()->stat_name_set_->add(name);
+  if (type == MetricType::Counter) {
+    auto id = wasm_->nextCounterMetricId();
+    auto c = &wasm()->scope_->counterFromStatName(stat_name);
+    wasm()->counters_.emplace(id, c);
+    *metric_id_ptr = id;
+    return WasmResult::Ok;
+  } else if (type == MetricType::Gauge) {
+    auto id = wasm()->nextGaugeMetricId();
+    auto g = &wasm()->scope_->gaugeFromStatName(stat_name, Stats::Gauge::ImportMode::Accumulate);
+    wasm()->gauges_.emplace(id, g);
+    *metric_id_ptr = id;
+    return WasmResult::Ok;
+  } else if (type == MetricType::Histogram) {
+    auto id = wasm()->nextHistogramMetricId();
+    auto h = &wasm()->scope_->histogramFromStatName(stat_name, Stats::Histogram::Unit::Unspecified);
+    wasm()->histograms_.emplace(id, h);
+    *metric_id_ptr = id;
+    return WasmResult::Ok;
+  }
+  return WasmResult::BadArgument;
+}
+
+WasmResult Context::incrementMetric(uint32_t metric_id, int64_t offset) {
+  auto type = static_cast<MetricType>(metric_id & Wasm::kMetricTypeMask);
+  if (type == MetricType::Counter) {
+    auto it = wasm()->counters_.find(metric_id);
+    if (it != wasm()->counters_.end()) {
+      if (offset > 0) {
+        it->second->add(offset);
+        return WasmResult::Ok;
+      } else {
+        return WasmResult::BadArgument;
+      }
+      return WasmResult::NotFound;
+    }
+  } else if (type == MetricType::Gauge) {
+    auto it = wasm()->gauges_.find(metric_id);
+    if (it != wasm()->gauges_.end()) {
+      if (offset > 0) {
+        it->second->add(offset);
+        return WasmResult::Ok;
+      } else {
+        it->second->sub(-offset);
+        return WasmResult::Ok;
+      }
+    }
+    return WasmResult::NotFound;
+  }
+  return WasmResult::BadArgument;
+}
+
+WasmResult Context::recordMetric(uint32_t metric_id, uint64_t value) {
+  auto type = static_cast<MetricType>(metric_id & Wasm::kMetricTypeMask);
+  if (type == MetricType::Counter) {
+    auto it = wasm()->counters_.find(metric_id);
+    if (it != wasm()->counters_.end()) {
+      it->second->add(value);
+      return WasmResult::Ok;
+    }
```

Comment:
It seems a little odd to me that record/increment do the same thing for counters? When doing these PRs can you please cross-link to the ABI definitions that we should be reviewing in parallel?

Reply:
@PiotrSikora can you link this here when the ABI doc is in the repo ?

Reply:
https://github.com/proxy-wasm/spec